File: packages/ruby-sdk/lib/opensend/resources/topics.rb
# frozen_string_literal: true

module OpenSend
  module Resources
    class Topics
      def initialize(http)
        @http = http
      end

      # POST /api/topics
      def create(params)
        @http.post("/api/topics", normalize(params))
      end

      # GET /api/topics
      def list(limit: nil, after: nil, search: nil)
        params = {}
        params[:limit] = limit unless limit.nil?
        params[:after] = after if after
        params[:search] = search if search
        @http.get("/api/topics", params)
      end

      # GET /api/topics/:id
      def get(id)
        @http.get("/api/topics/#{id}")
      end

      # PATCH /api/topics/:id
      def update(id, params)
        @http.patch("/api/topics/#{id}", normalize(params))
      end

      # DELETE /api/topics/:id
      def delete(id)
        @http.delete("/api/topics/#{id}")
      end

      private

      ALIASES = {
        "defaultSubscription" => "default_subscription"
      }.freeze

      def normalize(params)
        return {} if params.nil? || params.empty?

        result = {}
        params.each do |k, v|
          key = k.to_s
          result[ALIASES.fetch(key, key)] = v
        end
        result
      end
    end
  end
end

File: packages/go-sdk/topics.go
package opensend

import (
	"context"
	"fmt"
	"net/http"
)

// TopicsClient provides access to the /api/topics endpoints.
type TopicsClient struct {
	c *Client
}

// Create creates a new topic via POST /api/topics.
func (t *TopicsClient) Create(ctx context.Context, req CreateTopicRequest) (*CreateTopicResponse, error) {
	var out CreateTopicResponse
	if err := t.c.doJSON(ctx, http.MethodPost, "/api/topics", req, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// List retrieves a paginated list of topics via GET /api/topics.
func (t *TopicsClient) List(ctx context.Context, opts TopicListOptions) (*TopicListResponse, error) {
	params := map[string]string{}
	if opts.Limit != nil {
		params["limit"] = fmt.Sprintf("%d", *opts.Limit)
	}
	if opts.After != "" {
		params["after"] = opts.After
	}
	if opts.Search != "" {
		params["search"] = opts.Search
	}

	var out TopicListResponse
	if err := t.c.doJSON(ctx, http.MethodGet, buildQuery("/api/topics", params), nil, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get retrieves a single topic by ID via GET /api/topics/{id}.
func (t *TopicsClient) Get(ctx context.Context, id string) (*TopicResponse, error) {
	var out TopicResponse
	if err := t.c.doJSON(ctx, http.MethodGet, "/api/topics/"+id, nil, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Update partially updates a topic via PATCH /api/topics/{id}.
func (t *TopicsClient) Update(ctx context.Context, id string, req UpdateTopicRequest) (*TopicResponse, error) {
	var out TopicResponse
	if err := t.c.doJSON(ctx, http.MethodPatch, "/api/topics/"+id, req, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete removes a topic via DELETE /api/topics/{id}.
func (t *TopicsClient) Delete(ctx context.Context, id string) (*DeleteTopicResponse, error) {
	var out DeleteTopicResponse
	if err := t.c.doJSON(ctx, http.MethodDelete, "/api/topics/"+id, nil, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
